### PR TITLE
Add ProtocolStart and GracefulClose messages

### PR DESCRIPTION
**Pull request by gregtatcam**
## High Level Overview of Change

Add InboundHandoff class to send HTTP response 
to the outbound peer and instantiate inbound PeerImp
when the response is sent. The peer then waits for
mtSTART_PROTOCOL message from the connected 
outbound peer. When the message is received, the peer starts 
sending the protocol messages. This update ensures that 
the inbound peer is created once the overlay is established,
eliminating the need for doAccept(). In addition it synchronizes 
send/receive protocol messages between the peers on start,
eliminating the need for copying the read buffer into the outbound
peer constructor. 

Add mtGRACEFUL_CLOSE message to differentiate application
and link layer failures. 

### Type of Change

- [x]  Refactor (non-breaking change that only restructures code)

### CODE REVIEW

File: src/ripple/overlay/impl/InboundHandoff.cpp [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/overlay/impl/InboundHandoff.h>
```
Comment (HowardHinnant) [inline src/ripple/overlay/impl/InboundHandoff.cpp:20]: Needs additional includes to compile in non-unity build:
```c++
#include <ripple/overlay/Cluster.h>
#include <ripple/overlay/impl/PeerImp.h>
```

**Review by HowardHinnant — CHANGES_REQUESTED** [review]
ripple/overlay/impl/OverlayImpl.cpp needs an additional include to compile in a non-unity build:
```c++
#include <ripple/overlay/impl/InboundHandoff.h>
```

File: src/ripple/overlay/impl/InboundHandoff.h [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    boost::asio::strand<boost::asio::executor> strand_;
+    beast::IP::Endpoint const remote_address_;
+    ProtocolVersion protocol_;
+    PublicKey const publicKey_;
+    Resource::Consumer usage_;
+    std::shared_ptr<PeerFinder::Slot> const slot_;
+    http_request_type request_;
+
+public:
+    virtual ~InboundHandoff() = default;
```
Comment (HowardHinnant) [inline src/ripple/overlay/impl/InboundHandoff.h:53]: Suggest adding `override` qualifier to confirm that the base destructor is virtual.

File: src/ripple/overlay/impl/InboundHandoff.h [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
```
Comment (HowardHinnant) [inline src/ripple/overlay/impl/InboundHandoff.h:54]: Recommend adding:
```c++
   InboundHandoff(InboundHandoff const&) = delete;
   InboundHandoff& operator=(InboundHandoff const&) = delete;
 ```
To confirm and document that this class isn't meant to be copied.

**Review by HowardHinnant — APPROVED** [review]

**Comment by intelliot** [comment]
(notes: changes how servers communicate with each other; Nik will review later)

**Comment by intelliot** [comment]
@nbougalis you previously planned to review this; is that still the case?

**Comment by intelliot** [comment]
@gregtatcam do you think this PR still makes sense to merge in?

**Comment by gregtatcam** [comment]
> @gregtatcam do you think this PR still makes sense to merge in?

@nbougalis thought it was a useful enhancement. I don't have a strong opinion one way or another. There is certainly a risk since we are changing how the protocol starts/shuts down. This sequence had not changed in a very long time.

**Comment by ckeshava** [comment]
Apologies for the delay in my review. I had a few clarifying questions and I have posted the same.

I don't understand the limitations of the `doAccept()` function. Why are we trying to replace it? I get the performance benefits of the buffer-copy, but that seems to occur only at the beginning of the handshake right?

File: src/ripple/overlay/Message.h [hunk 4]
```diff
@@ -100,6 +100,14 @@ class Message : public std::enable_shared_from_this<Message>
         return validatorKey_;
     }
 
+    /** Get the message type from the payload header.
+     * First four bytes are the compression/algorithm flag and the payload size.
+     * Next two bytes are the message type
+     * @return Message type
+     */
+    int
+    getType() const;
```
[the remark this answers is not in the document]
Comment (gregtatcam) [reply src/ripple/overlay/Message.h:109]: It's `private` in the latest `develop` branch.

File: src/ripple/overlay/impl/Message.cpp [hunk 5]
```diff
@@ -221,9 +221,10 @@ Message::getBuffer(Compressed tryCompressed)
 }
 
 int
-Message::getType(std::uint8_t const* in) const
+Message::getType() const
 {
-    int type = (static_cast<int>(*(in + 4)) << 8) + *(in + 5);
+    int type =
```
[the remark this answers is not in the document]
Comment (gregtatcam) [reply src/ripple/overlay/impl/Message.cpp:226]: `setHeader()` comments:
>    All multi-byte values are represented in big endian.

    For uncompressed messages (6 bytes), numbering bits from left to right:

        - The first 6 bits are set to 0.
        - The next 26 bits represent the payload size.
        - The remaining 16 bits represent the message type.

    For compressed messages (10 bytes), numbering bits from left to right:

        - The first 32 bits, together, represent the compression algorithm
          and payload size:
            - The first bit is set to 1 to indicate the message is compressed.
            - The next 3 bits indicate the compression algorithm.
            - The next 2 bits are reserved at this time and set to 0.
            - The remaining 26 bits represent the payload size.
        - The next 16 bits represent the message type.
        - The remaining 32 bits are the uncompressed message size.
and this is how it's set 
>   *in++ = static_cast<std::uint8_t>((type >> 8) & 0xFF);
    *in++ = static_cast<std::uint8_t>(type & 0xFF);

[the remark this answers is not in the document]
Comment (ckeshava) [reply src/ripple/overlay/impl/Message.cpp:226]: ok, thanks for the explanation.

File: src/ripple/overlay/Message.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (ckeshava) [reply src/ripple/overlay/Message.h:109]: yes, I was referring to this feature branch. It is `public` in this branch, I'm wondering why it differs from the develop branch ?

[the remark this answers is not in the document]
Comment (gregtatcam) [reply src/ripple/overlay/Message.h:109]: To your point - it should had not be `public`. So it got fixed a while back.

[the remark this answers is not in the document]
Comment (ckeshava) [reply src/ripple/overlay/Message.h:109]: okay, thanks for the clarification.

**Comment by gregtatcam** [comment]
> Apologies for the delay in my review. I had a few clarifying questions and I have posted the same.
> 
> I don't understand the limitations of the `doAccept()` function. Why are we trying to replace it? I get the performance benefits of the buffer-copy, but that seems to occur only at the beginning of the handshake right?

It's not that `doAccept()` has limitation but the flow is not correct. At a super high level, this is how the overlay connection is established prior to this fix. 

The outbound connection (client) is handled by `ConnectAttempt`. `ConnectAttempt` sends the handshake once the connection is established. The handshake response is processed by `ConnectAttempt::processResponse()`, which verifies the handshake, instantiates outbound `PeerImp`, and adds it to `OverlayImpl` list of active peers (`OverlayImpl::add_active()`). This in turns runs the peer (`PeerImp::run()`) and for the outbound peer it starts the protocol (`doProtocolStart()`). The outbound peer is ready to process protocol messages at this point.

There is a lot more going on for the inbound connection. But ultimately it is handled by `OverlayImpl::onHandoff()`. `onHandoff()` verifies the handshake, instantiates the inbound peer, adds it to its list (not via `add_active()`), and runs the peer (`PeerImp::run()`). `run()` calls `doAccept()` for the inbound peer, which sends the handshake response, and then starts the protocol (`doProtocolStart()`). The inbound peer is ready to process protocol messages at this point. 

This flow sort of looks fine, but the problem is that the inbound peer is added to the overlay active list before the peer is ready to process the protocol messages; i.e. it has to send the handshake response first. This PR fixes this issue by adding `InboundHandoff` class, which is instantiated in `OverlayImpl::onHandoff()` instead of instantiating `PeerImp`. `InboundHandoff::sendResponse()` sends the handshake response. Once the response is written, `InboundHandoff` instantiates inbound `PeerImp` and adds it to `OverlayImpl` list of active peers (`OverlayImpl::add_active()`). Note how both inbound and outbound peer are added now consistently to the overlay active list via `add_active()`. This in turns runs the peer (`PeerImp::run()`) and starts the protocol (`doProtocolStart()`).  Again, this is handled consistently for both types of peers. Both inbound and outbound peers are active at this point. There is one more thing that needs to happen though. The inbound peer sent the response to the outbound peer but the inbound peer doesn't really know at this point if the outbound peer is ready. To kick start this process, the outbound peer sends `mtSTART_PROTOCOL` protocol message. Once the inbound peer receives this message, it starts sending the protocol messages. Everything is copacetic between the peers now.

This PR also adds `mtGRACEFUL_CLOSE` message to gracefully close the protocol whenever possible instead of just blindly closing the connection.

I have not looked at this code for two years so might have missed something...

**Comment by ckeshava** [comment]
thank you so much Greg, this explanation gave a lot of context to the PR change. 

In the current codebase, is it possible to have "phantom" peers? Since the inbound peer is added to the Overlay active peers list before receiving the handshake response, what is the fallback mechanism if we have incorrectly added a peer?

**Comment by gregtatcam** [comment]
> thank you so much Greg, this explanation gave a lot of context to the PR change.
> 
> In the current codebase, is it possible to have "phantom" peers? Since the inbound peer is added to the Overlay active peers list before receiving the handshake response, what is the fallback mechanism if we have incorrectly added a peer?

I don't think there is a fallback mechanism for this scenario.

**Comment by intelliot** [comment]
@ckeshava - what are your thoughts on this PR? Is it worth merging, or should it just be closed without merging?

**Review by ckeshava — APPROVED** [review]

**Comment by ckeshava** [comment]
I think this PR should be merged 👍

**Comment by intelliot** [comment]
@JoelKatz FYI - since this is an area of the code that hasn't been touched in many years, let me know if there is anything that needs to be further tested.

@sophiax851 it would be useful to confirm that none of these changes have adverse performance impact. Thank you!

**Comment by intelliot** [comment]
Thank you! This will be included in the next beta. If we find any issues with it, we can revert it as appropriate.

**Comment by intelliot** [comment]
note: we should check with @nbougalis on whether this should be kept in

**Pull request merged**